File: src-dotnet/api/endpoints/client/ClientEndpoint.cs
namespace MedrunnerApiClient.Api.Endpoints.Client
{
    /// <summary>
    /// Endpoints for interacting with clients.
    /// </summary>
    public class ClientEndpoint : ApiEndpoint
    {
        public ClientEndpoint(ApiConfig config, TokenManager tokenManager, ILogger logger)
            : base(config, tokenManager, logger)
        {
        }

        protected override string Endpoint => "client";

        /// <summary>
        /// Gets the current client.
        /// </summary>
        public async Task<ApiResponse<Models.Person>> GetAsync()
        {
            return await GetRequestAsync<Models.Person>("");
        }

        /// <summary>
        /// Gets the specified amount of emergencies the client has created.
        /// </summary>
        public async Task<ApiResponse<Api.PaginatedResponse<Models.ClientHistory>>> GetHistoryAsync(int limit, string? paginationToken = null)
        {
            var queryParams = new Dictionary<string, object> { { "limit", limit } };
            if (!string.IsNullOrEmpty(paginationToken))
                queryParams["paginationToken"] = paginationToken;
            return await GetRequestAsync<Api.PaginatedResponse<Models.ClientHistory>>("/history", queryParams);
        }

        /// <summary>
        /// Gets the blocklist status of the current client.
        /// </summary>
        public async Task<ApiResponse<string>> GetBlockedStatusAsync()
        {
            return await GetRequestAsync<string>("/blocked");
        }

        /// <summary>
        /// Links the current user to a rsiHandle.
        /// </summary>
        public async Task<ApiResponse<Models.Person>> LinkClientAsync(string rsiHandle)
        {
            var payload = new { rsiHandle };
            return await PostRequestAsync<Models.Person>("/link", payload);
        }

        /// <summary>
        /// Updates the settings of the current user for the Client Portal.
        /// </summary>
        public async Task<ApiResponse<string>> SetUserSettingsAsync(string settings)
        {
            var payload = new { settingsBlob = settings };
            return await PutRequestAsync<string>("/settings/clientPortal", payload);
        }

        /// <summary>
        /// Deactivate the current client.
        /// </summary>
        public async Task<ApiResponse<string>> DeactivateAsync()
        {
            return await DeleteRequestAsync<string>("");
        }
    }
}

File: src/api/endpoints/client/ClientEndpoint.ts
import { Logger } from "ts-log";

import { HeaderProvider } from "../../../Func";
import ClientHistory from "../../../models/ClientHistory";
import Person, { BlockedStatus } from "../../../models/Person";
import ApiResponse from "../../ApiResponse";
import PaginatedResponse from "../../PaginatedResponse";
import ApiEndpoint from "../ApiEndpoint";
import TokenManager from "../auth/TokenManager";
import DefaultApiConfig from "../DefaultApiConfig";

/**
 * Endpoints for interacting with clients.
 * */
export default class ClientEndpoint extends ApiEndpoint {
  constructor(config: DefaultApiConfig, tokenManager: TokenManager, log?: Logger, headerProvider?: HeaderProvider) {
    super(config, tokenManager, log, headerProvider);
  }

  protected override endpoint(): string {
    return "client";
  }

  /**
   * Gets the current client.
   * */
  public async get(): Promise<ApiResponse<Person>> {
    return await this.getRequest<Person>("");
  }

  /**
   * Gets the specified amount of emergencies the client has created.
   * @param limit - The number of emergencies to get, defaults to 10
   * @param paginationToken - The string to use for pagination
   * */
  public async getHistory(
    limit?: number,
    paginationToken?: string,
  ): Promise<ApiResponse<PaginatedResponse<ClientHistory>>> {
    return await this.getRequest<PaginatedResponse<ClientHistory>>("/history", { limit, paginationToken });
  }

  /**
   * Gets the blocklist status of the current client.
   * */
  public async getBlockedStatus(): Promise<ApiResponse<BlockedStatus>> {
    return await this.getRequest<BlockedStatus>("/blocked");
  }

  /**
   * Links the current user to a rsiHandle.
   *
   * @param rsiHandle - The RSI handle of the client
   * @returns The updated Person object of the client
   *
   * */
  public async linkClient(rsiHandle: string): Promise<ApiResponse<Person>> {
    return await this.postRequest<Person>("/link", { rsiHandle });
  }

  /**
   * Updates the settings of the current user for the Client Portal.
   *
   * @param settings - The stringified object settings to add or update
   *
   * */
  public async setUserSettings(settings: string): Promise<ApiResponse> {
    return await this.putRequest("/settings/clientPortal", { settingsBlob: settings });
  }

  /**
   * Deactivate the current client.
   * */
  public async deactivate(): Promise<ApiResponse> {
    return await this.deleteRequest("");
  }
}
